src/DependencyInjection/Compiler/BusBuilder/BusBuilder.php

```diff
@@ -0,0 +1,126 @@
+use Symfony\Component\DependencyInjection\ServiceLocator;
+
+final class BusBuilder
+{
+    /**
+     * @var string
+     */
+    private $busId;
+
+    /**
+     * @var string[]
+     */
+    private $middlewareIds = [];
+
+    /**
+     * @var string
+     */
+    private $methodInflectorId;
+
+    public function __construct(string $busId, string $methodInflector, array $middlewareIds)
+    {
+        $this->busId = $busId;
+        $this->methodInflectorId = $methodInflector;
+        $this->middlewareIds = $middlewareIds;
+    }
+
+    public function id(): string
+    {
+        return $this->busId;
+    }
+
+    public function serviceId(): string
+    {
+        return "tactician.commandbus.$this->busId";
+    }
+
+    public function locatorServiceId()
+    {
+        return "tactician.commandbus.{$this->busId}.handler.locator";
+    }
+
+    public function commandHandlerMiddlewareId(): string
+    {
+        return "tactician.commandbus.{$this->busId}.middleware.command_handler";
+    }
+
+    public function registerInContainer(ContainerBuilder $container, array $commandsToAccept)
+    {
+        $this->registerLocatorService($container, $commandsToAccept);
+
+        $container->setDefinition(
+            $this->commandHandlerMiddlewareId(),
+            new Definition(
+                CommandHandlerMiddleware::class,
+                [
+                    new Reference('tactician.handler.command_name_extractor.class_name'),
+                    new Reference($this->locatorServiceId()),
+                    new Reference($this->methodInflectorId),
+                ]
+            )
+        );
+
+        $container->setDefinition(
+            $this->serviceId(),
+            new Definition(
+                CommandBus::class,
+                [
+                    array_map(
+                        function (string $id) { return new Reference($id); },
+                        $this->middlewareIds
+                    )
+                ]
+            )
+        );
+    }
+
+    private function registerLocatorService(ContainerBuilder $container, $commandsToAccept)
+    {
+        // Leverage symfony/dependency-injection:^3.3 service locators
+        if (class_exists(ServiceLocator::class)) {
+            $definition = new Definition(
+                ContainerLocator::class,
+                [new Reference($this->registerHandlerServiceLocator($container, $commandsToAccept)), $commandsToAccept]
+            );
+        } else {
+            $definition = new Definition(
+                ContainerBasedHandlerLocator::class,
+                [new Reference('service_container'), $commandsToAccept]
+            );
+        }
+
+        $container->setDefinition($this->locatorServiceId(), $definition);
+    }
+
+    private function registerHandlerServiceLocator(ContainerBuilder $container, array $commandsToAccept): string
+    {
+        $handlers = [];
+        foreach ($commandsToAccept as $commandName => $id) {
```

Comment:
wdyt to rename `$id` into `$handlerId` to avoid any confusion ?

Reply:
Sure. 🤷‍